fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+//
+// The main loop of the fish program.
+/*
+Copyright (C) 2005-2008 Axel Liljencrantz
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License version 2 as
+published by the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
+*/
+
+use autocxx::prelude::*;
+
+use crate::{
+    ast::Ast,
+    builtins::shared::{
+        BUILTIN_ERR_MISSING, BUILTIN_ERR_UNKNOWN, STATUS_CMD_OK, STATUS_CMD_UNKNOWN,
+    },
+    common::{
+        escape_string, exit_without_destructors, get_executable_path, str2wcstring, wcs2string,
+        EscapeStringStyle, PROFILING_ACTIVE, PROGRAM_NAME,
+    },
+    env::{
+        environment::{env_init, EnvStack, Environment},
+        ConfigPaths, EnvMode,
+    },
+    event::{self, Event},
+    ffi::{self, Repin},
+    flog::{self, activate_flog_categories_by_pattern, set_flog_file_fd, FLOG, FLOGF},
+    function, future_feature_flags as features,
+    history::start_private_mode,
+    parse_constants::{ParseErrorList, ParseErrorListFfi, ParseTreeFlags},
+    parse_tree::{ParsedSource, ParsedSourceRefFFI},
+    path::path_get_config,
+    signal::{signal_clear_cancel, signal_unblock_all},
+    threads::{self, asan_maybe_exit},
+    topic_monitor,
+    wchar::prelude::*,
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+    wutil::waccess,
+};
+use std::env;
+use std::ffi::{CString, OsStr, OsString};
+use std::fs::File;
+use std::io;
+use std::mem::MaybeUninit;
+use std::os::unix::prelude::*;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
```

Comment:
> 3.6.1-99999-dirty

Ideally the commit that fixes this comes before this one (to minimize eyebrow-raising)